<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="stampen">
<title>`stampen` Usage Vignette • stampen</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="`stampen` Usage Vignette">
<meta property="og:description" content="stampen">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">stampen</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/simulation_vignette.html">`stampen` Usage Vignette</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>`stampen` Usage Vignette</h1>
                        <h4 data-toc-skip class="author">Jonah
Einson</h4>
            
            <h4 data-toc-skip class="date">1/5/23</h4>
      
      
      <div class="d-none name"><code>simulation_vignette.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jeinson.github.io/stampen/">stampen</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<p><code>stampen</code> is a package used to simulate haplotype
configuration data from large phased WGS cohorts, and test for
non-random depletion of QTL-coding variant haplotypes using the
Poisson-binomial distribution. This non-random depletion can be
indicative of a selective effect which favors haplotypes in
non-deleterious configurations. This phenomenon is known as “modified
penetrance.”</p>
<div class="section level2">
<h2 id="tldr">TLDR<a class="anchor" aria-label="anchor" href="#tldr"></a>
</h2>
<p>A very quick overview of this tool’s functionality, for the lazy.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate some null haplotypes</span></span>
<span><span class="va">test_data</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/simulate_null_haplotypes.html">simulate_null_haplotypes</a></span><span class="op">(</span></span>
<span>    n_indvs <span class="op">=</span> <span class="fl">500</span>, n_genes <span class="op">=</span> <span class="fl">500</span>, </span>
<span>    qtl_alpha <span class="op">=</span> <span class="fl">1</span>, qtl_beta <span class="op">=</span> <span class="fl">1</span> <span class="co"># controls the qtl allele frequency distribution</span></span>
<span>    <span class="co"># based on a Beta</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Characterize the haplotypes as high or low penetrance</span></span>
<span><span class="va">test_data_incl_beta</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/characterize_haplotypes.html">characterize_haplotypes</a></span><span class="op">(</span></span>
<span>    <span class="va">test_data</span>, beta_config <span class="op">=</span> <span class="va">beta_config_sqtl</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">test_data_incl_beta</span><span class="op">)</span></span>
<span><span class="co">#&gt;    gene    indv haplotype sqtl_af beta   hom  exp_beta</span></span>
<span><span class="co">#&gt; 1 gene5 indv245      AbaB   0.773    0 FALSE 0.5000000</span></span>
<span><span class="co">#&gt; 2 gene6 indv228      ABAb   0.674    0  TRUE 0.8104083</span></span>
<span><span class="co">#&gt; 3 gene6 indv241      ABab   0.674    1 FALSE 0.5000000</span></span>
<span><span class="co">#&gt; 4 gene6 indv246      ABAb   0.674    0  TRUE 0.8104083</span></span>
<span><span class="co">#&gt; 5 gene6 indv309      ABAb   0.674    0  TRUE 0.8104083</span></span>
<span><span class="co">#&gt; 6 gene6 indv375      ABAb   0.674    0  TRUE 0.8104083</span></span>
<span></span>
<span><span class="co"># Run the Poisson-binomial bootstrap test</span></span>
<span><span class="fu"><a href="../reference/bootstrap_test.html">bootstrap_test</a></span><span class="op">(</span><span class="va">test_data_incl_beta</span>, B <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="co">#&gt;   bootstrap_p       epsilon         lower         upper  n_haplotypes </span></span>
<span><span class="co">#&gt;   0.922000000   0.002546359  -0.039555247   0.049623976 725.000000000</span></span></code></pre></div>
<div class="section level3">
<h3 id="simulating-data">Simulating Data<a class="anchor" aria-label="anchor" href="#simulating-data"></a>
</h3>
<p>In this vignette, we demonstrate how to simulate some haplotype data
with the <code>stampen</code> package, then test to see if is enriched
for haplotypes that reduce variant penetrance.</p>
<p>First, we simulate a dataset of haplotypes with from 500 individuals,
and 500 genes that may carry rare variants. The
<code>simulate_null_haplotypes</code> function generates data where
haplotypes in low penetrance and high penetrance configurations are
equally likely to occur. By default, we draw allele frequencies from a
uniform distribution, but this can be adjusted by setting the
<code>qtl_alpha</code> and <code>qtl_beta</code>parameters.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234567890</span><span class="op">)</span></span>
<span><span class="va">test_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_null_haplotypes.html">simulate_null_haplotypes</a></span><span class="op">(</span></span>
<span>  n_indvs <span class="op">=</span> <span class="fl">500</span>, n_genes <span class="op">=</span> <span class="fl">500</span>, </span>
<span>  qtl_alpha <span class="op">=</span> <span class="fl">1</span>, qtl_beta <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="co"># leave the default hyperparameters</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">test_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;    gene    indv haplotype qtl_af</span></span>
<span><span class="co">#&gt; 1 gene1 indv462      abaB  0.135</span></span>
<span><span class="co">#&gt; 2 gene2 indv126      ABab  0.187</span></span>
<span><span class="co">#&gt; 3 gene2 indv177      aBab  0.187</span></span>
<span><span class="co">#&gt; 4 gene2 indv215      aBab  0.187</span></span>
<span><span class="co">#&gt; 5 gene2 indv268      abaB  0.187</span></span>
<span><span class="co">#&gt; 6 gene2 indv339      abaB  0.187</span></span></code></pre></div>
<p>The <code>haplotype</code> column will be discussed further down.</p>
<p>In the default configuration, the QTL allele frequency is
uniform.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">gene</span>, <span class="va">qtl_af</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="va">unique</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">$</span><span class="va">qtl_af</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span>main <span class="op">=</span> <span class="st">"Histogram simulated QTL allele frequencies"</span>, </span>
<span>       xlab <span class="op">=</span> <span class="st">"QTL Af"</span>, </span>
<span>       col <span class="op">=</span> <span class="st">"cornflowerblue"</span><span class="op">)</span></span></code></pre></div>
<p><img src="simulation_vignette_files/figure-html/unnamed-chunk-6-1.png" width="576"></p>
<p>We can also adjust the beta distribution parameters to simulate data
that arises from any allele frequency distribution that may occur among
different sets of QTLs, or populations.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use this built in function to quickly visualize different theoretical allele </span></span>
<span><span class="co"># frequency distributions, paramterized by </span></span>
<span><span class="fu">stampen</span><span class="fu">:::</span><span class="fu"><a href="../reference/beta_plotter.html">beta_plotter</a></span><span class="op">(</span>a <span class="op">=</span> <span class="fl">2</span>, b <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="simulation_vignette_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">128</span><span class="op">)</span></span>
<span><span class="va">test_data_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_null_haplotypes.html">simulate_null_haplotypes</a></span><span class="op">(</span>n_indvs <span class="op">=</span> <span class="fl">500</span>, n_genes <span class="op">=</span> <span class="fl">500</span>, </span>
<span>                                        qtl_alpha <span class="op">=</span> <span class="fl">2</span>, qtl_beta <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="co"># new hps</span></span>
<span></span>
<span><span class="va">test_data_2</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">gene</span>, <span class="va">qtl_af</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="va">unique</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">$</span><span class="va">qtl_af</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span>main <span class="op">=</span> <span class="st">"Histogram simulated QTL allele frequencies"</span>, </span>
<span>       xlab <span class="op">=</span> <span class="st">"QTL Af"</span>, </span>
<span>       col <span class="op">=</span> <span class="st">"firebrick"</span><span class="op">)</span></span></code></pre></div>
<p><img src="simulation_vignette_files/figure-html/unnamed-chunk-8-1.png" width="576"></p>
</div>
<div class="section level3">
<h3 id="haplotype-configuration-designations">Haplotype configuration designations<a class="anchor" aria-label="anchor" href="#haplotype-configuration-designations"></a>
</h3>
<p>Let’s take a look at the built in haplotype configurations, which we
consider “high-penetrance” or “low-penetrance.” In a haplotype
configuration analysis, an important first step is to hypothesize
whether higher expression of a nonmutated allele, or lower inclusion of
a mutated allele is what drives modified penetrance. Testing for either
is a decision one needs to make. These designations can be adjusted
easily by the user.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">'beta_config_sqtl'</span><span class="op">)</span></span>
<span><span class="va">beta_config_sqtl</span></span>
<span><span class="co">#&gt; abAB ABab abaB aBab AbaB aBAb AbAB ABAb </span></span>
<span><span class="co">#&gt;    1    1    1    1    0    0    0    0</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">'beta_config_eqtl'</span><span class="op">)</span></span>
<span><span class="va">beta_config_eqtl</span></span>
<span><span class="co">#&gt; abAB ABab abaB aBab AbaB aBAb AbAB ABAb </span></span>
<span><span class="co">#&gt;    0    0    1    1    1    1    0    0</span></span></code></pre></div>
<p>In the sQTL hypothesis configuation, <code>a</code> represents a high
exon inclusion alelle, and <code>A</code> represents a low inclusion
allele. (In the eQTL model, <code>A</code> is a highly expressed allele,
and <code>a</code> is the lowly expressed allele). Therefore, under the
sQTL model, an observation of the haplotype <code>abAB</code> means an
individual is heteryzous for an sQTL, and carries a rare coding variant
on the highly included allelic copy. This configuration is assigned a
<code>1</code> becuase under this model, this configuration results in
increased penetrance of the coding variant, with respect to the opposite
configuration <code>AbaB</code>, where the coding variant is now on the
lower included allele, and has reduced penetrance. (See figure 1 of
Einson et. al. 2023)</p>
<p>Next, we use the function <code>characterize_haplotypes</code> to add
two columns to a data frame of haplotypes. The new columns are.</p>
<ul>
<li>
<strong>beta</strong>: Represents a high penetrance (1) or low
penetrance (0) configuration, which is defined by the input
<code>beta_config</code>.</li>
<li>
<strong>hom</strong>: Boolean. Is this individual homozygous for the
QTL allele?</li>
<li>
<strong>exp_beta</strong>: When hom == <code>TRUE</code>, this is
calculated as <code>sqtl_af^2 / (sqtl_af^2 + (1 - sqtl_af^2))</code>.
When <code>hom</code> == <code>TRUE</code>, i.e. the haplotype is
hetorozygous, we assign 0.5. In practice, <code>sqtl_af</code> is the
allele frequency of the <em>penetrance driving allele</em>. For sQTLs,
this is the frequency of the high exon inclusion alelle, and for eQTLs,
this is the allele frequency of the lower expressed allele.</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_data_incl_beta</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/characterize_haplotypes.html">characterize_haplotypes</a></span><span class="op">(</span><span class="va">test_data</span>, beta_config <span class="op">=</span> <span class="va">beta_config_sqtl</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">test_data_incl_beta</span>, <span class="fl">20</span><span class="op">)</span></span>
<span><span class="co">#&gt;      gene    indv haplotype sqtl_af beta   hom   exp_beta</span></span>
<span><span class="co">#&gt; 1   gene1 indv462      abaB   0.135    1  TRUE 0.02377846</span></span>
<span><span class="co">#&gt; 2   gene2 indv126      ABab   0.187    1 FALSE 0.50000000</span></span>
<span><span class="co">#&gt; 3   gene2 indv177      aBab   0.187    1  TRUE 0.05024729</span></span>
<span><span class="co">#&gt; 4   gene2 indv215      aBab   0.187    1  TRUE 0.05024729</span></span>
<span><span class="co">#&gt; 5   gene2 indv268      abaB   0.187    1  TRUE 0.05024729</span></span>
<span><span class="co">#&gt; 6   gene2 indv339      abaB   0.187    1  TRUE 0.05024729</span></span>
<span><span class="co">#&gt; 7   gene2 indv345      abAB   0.187    1 FALSE 0.50000000</span></span>
<span><span class="co">#&gt; 8   gene2 indv484      abaB   0.187    1  TRUE 0.05024729</span></span>
<span><span class="co">#&gt; 9   gene5  indv91      abaB   0.064    1  TRUE 0.00465353</span></span>
<span><span class="co">#&gt; 10  gene5 indv206      abaB   0.064    1  TRUE 0.00465353</span></span>
<span><span class="co">#&gt; 11  gene5 indv269      aBab   0.064    1  TRUE 0.00465353</span></span>
<span><span class="co">#&gt; 12  gene5 indv366      aBab   0.064    1  TRUE 0.00465353</span></span>
<span><span class="co">#&gt; 13  gene6 indv465      aBab   0.117    1  TRUE 0.01725407</span></span>
<span><span class="co">#&gt; 14  gene8  indv69      aBab   0.341    1  TRUE 0.21120419</span></span>
<span><span class="co">#&gt; 15  gene8 indv152      aBab   0.341    1  TRUE 0.21120419</span></span>
<span><span class="co">#&gt; 16  gene8 indv208      ABab   0.341    1 FALSE 0.50000000</span></span>
<span><span class="co">#&gt; 17  gene8 indv287      ABab   0.341    1 FALSE 0.50000000</span></span>
<span><span class="co">#&gt; 18  gene8 indv293      abAB   0.341    1 FALSE 0.50000000</span></span>
<span><span class="co">#&gt; 19  gene8 indv480      abaB   0.341    1  TRUE 0.21120419</span></span>
<span><span class="co">#&gt; 20 gene10   indv7      ABAb   0.721    0  TRUE 0.86976185</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="test-for-depletion-of-high-penetrance-haplotypes">Test for depletion of high penetrance haplotypes<a class="anchor" aria-label="anchor" href="#test-for-depletion-of-high-penetrance-haplotypes"></a>
</h3>
<p>Now use the <code>calculate_epsilon</code> and
<code>bootstrap_test</code> functions to calculate depletion, which we
call epsilon, and get the significance of this depletion.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/calculate_epsilon.html">calculate_epsilon</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="va">test_data_incl_beta</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] -0.008338446</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootstrap_test.html">bootstrap_test</a></span><span class="op">(</span><span class="va">test_data_incl_beta</span>, B <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">test_res</span></span>
<span><span class="co">#&gt;   bootstrap_p       epsilon         lower         upper  n_haplotypes </span></span>
<span><span class="co">#&gt;   0.720000000  -0.008338446  -0.051953185   0.035550769 773.000000000</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Make a nice figure</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">test_res</span><span class="op">)</span><span class="op">)</span>, </span>
<span>       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">epsilon</span>, x <span class="op">=</span> <span class="st">"baseline"</span>, ymin <span class="op">=</span> <span class="va">lower</span>, ymax <span class="op">=</span> <span class="va">upper</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html" class="external-link">geom_pointrange</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, lty <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="op">-</span><span class="fl">.1</span>, <span class="fl">.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"ε"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.title.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.ticks.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.line.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="simulation_vignette_files/figure-html/unnamed-chunk-14-1.png" width="576"></p>
<p>Now as a demonstration, let’s remove some “high penetrance”
haplotypes to show how sensitive the test is for detecing this
signal.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">test_data_incl_beta</span><span class="op">$</span><span class="va">beta</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">idx.removed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">idx</span>, <span class="fl">.15</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="va">test_data_incl_beta_depleted</span> <span class="op">&lt;-</span> <span class="va">test_data_incl_beta</span><span class="op">[</span><span class="op">-</span><span class="va">idx.removed</span>,<span class="op">]</span></span>
<span><span class="va">test_res_depl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootstrap_test.html">bootstrap_test</a></span><span class="op">(</span><span class="va">test_data_incl_beta_depleted</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">test_res</span>, <span class="va">test_res_depl</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">plt</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="va">plt</span><span class="op">)</span></span>
<span><span class="va">n_pos</span> <span class="op">=</span> <span class="fl">.2</span>; <span class="va">pval_pos</span> <span class="op">=</span> <span class="fl">.15</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plt</span>,</span>
<span>       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">epsilon</span>, x <span class="op">=</span> <span class="st">"baseline"</span>, col <span class="op">=</span> <span class="va">rowname</span>, ymin <span class="op">=</span> <span class="va">lower</span>, ymax <span class="op">=</span> <span class="va">upper</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html" class="external-link">geom_pointrange</a></span><span class="op">(</span>position <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_dodge.html" class="external-link">position_dodge</a></span><span class="op">(</span><span class="fl">.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, lty <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="op">-</span><span class="fl">.2</span>, <span class="fl">.2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"ε"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># The P-value label</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">pval_pos</span>, label <span class="op">=</span> <span class="va">bootstrap_p</span><span class="op">)</span>,</span>
<span>            position <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_dodge.html" class="external-link">position_dodge</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  </span>
<span>  <span class="co"># The N-sample label</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">n_pos</span>, label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/formatc.html" class="external-link">prettyNum</a></span><span class="op">(</span><span class="va">n_haplotypes</span>, big.mark <span class="op">=</span> <span class="st">","</span><span class="op">)</span><span class="op">)</span>,</span>
<span>            position <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_dodge.html" class="external-link">position_dodge</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.title.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.ticks.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.line.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="simulation_vignette_files/figure-html/unnamed-chunk-16-1.png" width="576"></p>
</div>
<div class="section level3">
<h3 id="run-the-tutorial">Run the tutorial<a class="anchor" aria-label="anchor" href="#run-the-tutorial"></a>
</h3>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate_epsilon(test_data, beta_config = beta_config_sqtl)</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Jonah Einson, Pejman Mohammadi.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
